### DISCUSSION TITLE
What about a possibility for a driver to set PKCS#11 token model ?

### DISCUSSION THREAD

**Discussion by llogar** [OpenSC/OpenSC#2909] [Ideas]
Currently PKCS#11 token model is either read from the conf file or set to generic `PKCS#15` or `PKCS#15 emulated` values if `aplication.model` is not present in conf file.

I think it would be nice if the card driver could set the model via the `sc_card_ctl(...)` call with a fallback to legacy approach?
This way new models can be introduced without touching the conf file and/or be changed on the fly (or as needed)?

I have already made necessary changes and can provide a PR if you think this could be usable?

**Comment by Jakuje** [comment]
I agree that the `PKCS#15 emulated` is not much helpful for the user. I think setting model from configuration file is not much helpful either.

The PKCS#11 specs does not go into detail what specific values should be in this field:

> **model**       model of the device.  MUST be padded with the blank character (‘ ‘).  MUST NOT be null-terminated.

https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/pkcs11-base-v3.0.html

So I would say that as long as it can provide some useful information to the user, I think it would be good to have. I am not sure though if it should be another `sc_card_ctl()`, or we should take similar approach as we have for label, which is taken from existing variable set by driver earlier.

I think the card driver is the part that has the most information about the card (usually based on ATR) so it should be the one setting the model.

So my take would be to go ahead and open a PR. The rest can be discussed during reviews.
